### ISSUE TITLE
weird shadowing behavior regarding generic type names

### ISSUE THREAD

**Issue by lenscas** [teal-language/tl#764]
in this [playground](https://teal-playground.netlify.app/?c=bG9jYWwgcmVjb3JkIHRlc3QKICAgIHJlY29yZCBGb288aW50ZWdlcj4gLS10aGlzIHNob3VsZCBlaXRoZXIgZXJyb3IsIG9yIG90aGVyd2lzZSBzaGFkb3cgYGludGVnZXJgCiAgICAgICAgdXNlcmRhdGEKICAgICAgICAtLWluc3RlYWQsIGl0IGlzIGFsbG93ZWQgX2J1dF8gYWxzbyBkb2Vzbid0IHNoYWRvdyBpdC4gTWFraW5nIHRoZSBpbnRlZ2VyIGhlcmUgcmVmZXIgdG8KICAgICAgICAtLXRoZSBhY3R1YWwgaW50ZWdlciB0eXBlLCBhbmQgbm90IHRoZSBnZW5lcmljIHR5cGUgbmFtZQogICAgICAgIGJhciA6IGZ1bmN0aW9uKHNlbGY6Rm9vPGludGVnZXI%2BKTooc3RyaW5nKQogICAgICAgIC0tc2FtZSBkZWFsIGhlcmUuIFRoaXMgc2hvdWxkJ3ZlIGVycm9yZWQgb3Igc2hhZG93ZWQgYm90aCB0aGUgZ2VuZXJpYyBhbmQgdGhlIHJlYWwgaW50ZWdlciB0eXBlCiAgICAgICAgLS1idXQuLi4gaXQgZG9lcyBuZXZlcgogICAgICAgIGJhcjI6IGZ1bmN0aW9uPGludGVnZXI%2BKHNlbGY6Rm9vPGludGVnZXI%2BKTooc3RyaW5nKQogICAgZW5kCmVuZAoKbG9jYWwgeDogdGVzdC5Gb288bnVtYmVyPjsgLS1wcmV0ZW5kIHRoYXQgWCBpcyB2YWxpZC4uLgp4OmJhcigpIC0tYmFyIGlzIGludmFsaWQgdG8gY2FsbCwgYmVjYXVzZSBpdCBleHBlY3RzIGEgRm9vPGludGVyZ2VyPiBidXQgd2UgaGF2ZSBhIEZvbzxudW1iZXI%2BLi4uCng6YmFyMigpIC0tYmFyMiBpcyBhbHNvIGludmFsaWQgYXMgaXQgcmVhbGx5IHdhbnRzIGEgRm9vPGludGVnZXI%2BLi4u) 
teal at first accepts the name `integer` as the name of the generic for Foo, but when then using this instead reads this as the built in `integer` type.

Ideally this should either throw an error or properly shadow the `integer` type. 

This becomes even more confusing for people when instead of a built in type like `integer` we instead get a custom type involved like on this [playground](https://teal-playground.netlify.app/?c=bG9jYWwgcmVjb3JkIEZvb3MKZW5kCmxvY2FsIHJlY29yZCB0ZXN0CiAgICByZWNvcmQgRm9vPEZvb3M%2BIC0tdGhpcyBzaG91bGQgZWl0aGVyIGVycm9yLCBvciBvdGhlcndpc2Ugc2hhZG93IGBpbnRlZ2VyYAogICAgICAgIHVzZXJkYXRhCiAgICAgICAgLS1pbnN0ZWFkLCBpdCBpcyBhbGxvd2VkIF9idXRfIGFsc28gZG9lc24ndCBzaGFkb3cgaXQuIE1ha2luZyB0aGUgaW50ZWdlciBoZXJlIHJlZmVyIHRvCiAgICAgICAgLS10aGUgYWN0dWFsIGludGVnZXIgdHlwZSwgYW5kIG5vdCB0aGUgZ2VuZXJpYyB0eXBlIG5hbWUKICAgICAgICBiYXIgOiBmdW5jdGlvbihzZWxmOkZvbzxGb29zPik6KHN0cmluZykKICAgICAgICAtLXNhbWUgZGVhbCBoZXJlLiBUaGlzIHNob3VsZCd2ZSBlcnJvcmVkIG9yIHNoYWRvd2VkIGJvdGggdGhlIGdlbmVyaWMgYW5kIHRoZSByZWFsIGludGVnZXIgdHlwZQogICAgICAgIC0tYnV0Li4uIGl0IGRvZXMgbmV2ZXIKICAgICAgICBiYXIyOiBmdW5jdGlvbjxGb29zPihzZWxmOkZvbzxGb29zPik6KHN0cmluZykKICAgIGVuZAplbmQKCmxvY2FsIHg6IHRlc3QuRm9vPG51bWJlcj47IC0tcHJldGVuZCB0aGF0IFggaXMgdmFsaWQuLi4KeDpiYXIoKSAtLWJhciBpcyBpbnZhbGlkIHRvIGNhbGwsIGJlY2F1c2UgaXQgZXhwZWN0cyBhIEZvbzxpbnRlcmdlcj4gYnV0IHdlIGhhdmUgYSBGb288bnVtYmVyPi4uLgp4OmJhcjIoKSAtLWJhcjIgaXMgYWxzbyBpbnZhbGlkIGFzIGl0IHJlYWxseSB3YW50cyBhIEZvbzxpbnRlZ2VyPi4uLg%3D%3D)

Because here the generic typename correctly shadows the custom type and everything ends up behaving as expected.

**Comment by hishamhm** [comment]
Agreed that it needs to emit at the very least a redeclaration warning, and handling of builtin type names where type identifiers are declared everywhere. This for example, behaves weirdly too:

```lua
local record integer -- huh, no error
   x: string
end

local r: integer = { x = "hello" } -- cannot assign a table to an integer
```

**Comment by hishamhm** [comment]
Fixed in `next`! See [1035fd7](https://github.com/teal-language/tl/commit/1035fd7f18e75d37106092dea053296a57e96c84)
